### PR TITLE
Enable retrying flaky tests in nightly CI

### PR DESCRIPTION
**Pull request by neilmayhew**
Nightly tests continue to fail randomly due to an excessive number of discards. Fixing the discards problem will be difficult, but we can work around it for now by retrying test suites that have failed for this specific reason. However, we do this only for nightly CI and not for other workflow runs.

### CODE REVIEW

File: .github/workflows/haskell.yml
Comment (lehins) [inline .github/workflows/haskell.yml]: We don't need to force people reading that horrible ugly bash script if they don't care about nightly tests.
```suggestion
        if [ -z "$NIGHTLY" ]; then
          cabal test ${{ matrix.package }}
        else
          TRIES=2
          scripts/cabal-test-with-retries.sh "${{ matrix.package }}" "$TRIES"
        fi         
```

**Review by lehins — APPROVED** [review]
This approach might result in longer nightly runs without really fixing the problem (no guarantee that 2 or 3 or even 5 retries is enough for nightly tests to succeed)

That being said, we can give it a try, since it only affects the nightly runs there is no danger in this.

File: .github/workflows/haskell.yml
Comment (neilmayhew) [reply .github/workflows/haskell.yml]: Yes, that seems better. I was going to do it that way originally but thought it would be more messy in the workflow file than it actually is, and the script ended up being more messy than I expected :-)

**Comment by neilmayhew** [comment]
> This approach might result in longer nightly runs without really fixing the problem (no guarantee that 2 or 3 or even 5 retries is enough for nightly tests to succeed)

From what I've observed, the individual test suites fail only occasionally, but when you're running 108 of them the chance of at least one failing is quite high. Rerunning just the failed ones seems to have a good chance of succeeding. (I tried that manually on last night's CI before starting this.)

Most of the the individual test suites are fairly quick to execute, so if only a couple are failing it doesn't add much additional time to run them twice. In fact, since there's a lot of suites running in parallel, running a shorter one twice might still finish before the longer ones.

File: .github/workflows/haskell.yml
Comment (neilmayhew) [reply .github/workflows/haskell.yml]: I also simplified the script now that it doesn't have to handle the `TRIES=1` case cleanly

**Comment by MaximilianAlgehed** [comment]
FYI I took the time to look into what it would take to update the QuickCheck dependency to 2.15 to fix the flaky test in shelley and these are the packages for which you'd need `allow-newer`:
```
  ral:QuickCheck,
  fin:QuickCheck,
  bin:QuickCheck,
  tree-diff:QuickCheck,
  base64-bytestring-type:QuickCheck
```
The thing about this is that the only breaking change in the latest version of QuickCheck is that properties that don't do any random generation are no longer implicitly tested only once. This shouldn't break any of these packages in any significant way (as we aren't using any properties from those packages anyway) so it's up to @lehins if you think it's an acceptable trade-off to introduce these to `cabal.project` to (hopefully) fix the flaky test issue. However, I understand if `allow-newer` makes people queezy!

The QuickCheck bump is on branch https://github.com/IntersectMBO/cardano-ledger/tree/bump-quickcheck but when doing this I discovered a very stupid bug that I introduced in QuickCheck so we're going to have to wait for 2.15.1 anyway. That gives us some more time to wait for the maintainers of the above packages to catch up.

**Comment by MaximilianAlgehed** [comment]
(the QuickCheck bump is on branch https://github.com/IntersectMBO/cardano-ledger/tree/bump-quickcheck)

**Comment by lehins** [comment]
> it's up to @lehins if you think it's an acceptable trade-off to introduce these to cabal.project to (hopefully) fix the flaky test issue. However, I understand if allow-newer makes people queezy!

`allow-newer` would fix it for us, but then it could confuse ledger users into believing that they also need the newer version of QuickCheck. 
I don't think there is any rush on this, eventually we will all get onto the new version of QuickCheck and this flakiness issue will go away. In other words IMHO there is no point in putting effort and causing disturbance for people if the issue will resolve itself with time, especially since this interim solution should unblock the nightly CI and recent improvements to CI make dealing with flaky tests a lot less painfull.

**Pull request merged**
